src/algorithms/signal_source/gnuradio_blocks/fifo_reader.cc

```diff
@@ -80,35 +88,102 @@ int FifoReader::work(int noutput_items,
     return this->WORK_CALLED_PRODUCE;
 }
 
+
+template <typename Type>
+size_t FifoReader::read_interleaved(int noutput_items, gr_vector_void_star &output_items)
+{
+    boost::mutex::scoped_lock lock(d_mutex);  // hold mutex for duration of this function
+    size_t items_retrieved = 0;
+    int n;
+    Type real;
+    Type imag;
+
+    if (buffer_idx_ >= buffer_size_)
+        {
+            fifo_.read(fifo_buffer_, FIFO_SIZE);
+            buffer_idx_ = 0;
+            if (fifo_.good())
+                {
+                    buffer_size_ = FIFO_SIZE;
+                }
+            else if (fifo_.eof())
+                {
+                    // Although we got an EOF, ensure we don't lose the other samples
+                    buffer_size_ = fifo_.gcount();
+                    fifo_.clear();
+                }
+            else
+                {
+                    fifo_error_output();
+                    return 0;
+                }
+        }
+
+    for (n = 0; n < noutput_items && buffer_idx_ < buffer_size_; n++)
+        {
+            memcpy(&real, &fifo_buffer_[buffer_idx_], sizeof(real));
+            memcpy(&imag, &fifo_buffer_[buffer_idx_ + sizeof(imag)], sizeof(imag));
+            static_cast<gr_complex *>(output_items.at(0))[n] = gr_complex(real, imag);
+
+            buffer_idx_ += 2 * sizeof(Type);
+        }
+
+    items_retrieved = n;
+
+    return items_retrieved;
+}
+
+
 // read gr_complex items from fifo
-// this fct has duplicate code with the templated read_interleaved fct in header
+// this fct has duplicate code with the templated read_interleaved fct above
 size_t FifoReader::read_gr_complex(int noutput_items, gr_vector_void_star &output_items)
 {
+    boost::mutex::scoped_lock lock(d_mutex);  // hold mutex for duration of this function
     size_t items_retrieved = 0;
-    for (int n = 0; n < noutput_items; n++)
+    int n;
+    gr_complex sample;
+
+    if (buffer_idx_ >= buffer_size_)
         {
-            gr_complex sample;
-            fifo_.read(reinterpret_cast<char *>(&sample), sizeof(sample));
+            fifo_.read(fifo_buffer_, FIFO_SIZE);
```

Comment:
Note that the FIFO is buffered (`ifstream`) so you aren't gaining anything by buffering yourself here. If you have empirical evidence that reading one item at a time is "too slow" and that this buffering is much faster, then you owe future readers a comment to that effect. However, I'm hard pressed to see how this buffering -> single item extraction can be faster than the standard library.